api/circulation.py

```diff
@@ -966,17 +967,36 @@ def fulfill(self, patron, pin, licensepool, delivery_mechanism, part=None, fulfi
             if licensepool.self_hosted:
                 fulfillment = self._try_to_sign_fulfillment_link(licensepool, fulfillment)
         else:
-            api = self.api_for_license_pool(licensepool)
-            internal_format = api.internal_format(delivery_mechanism)
-
-            # Here we _do_ pass in the vendor-specific arguments, but
-            # we pass them in as keyword arguments, to minimize the
-            # impact on implementation signatures. Most vendor APIs
-            # will ignore one or more of these arguments.
-            fulfillment = api.fulfill(
-                patron, pin, licensepool, internal_format=internal_format,
-                part=part, fulfill_part_url=fulfill_part_url
-            )
+            if loan and loan.cached_manifest:
+                # Skip the fulfill call and build out FulfillmentInfo
+                # using information from the Loan
+                fulfillment = FulfillmentInfo(
+                    collection=licensepool.collection,
+                    data_source_name=licensepool.data_source,
+                    identifier_type=licensepool.identifier.type,
+                    identifier=licensepool.identifier.identifier,
+                    content_link=None,
+                    content_type=loan.cached_content_type,
+                    content=loan.cached_manifest,
+                    content_expires=loan.end
+                )
+            else:
+                api = self.api_for_license_pool(licensepool)
+                internal_format = api.internal_format(delivery_mechanism)
+
+                # Here we _do_ pass in the vendor-specific arguments, but
+                # we pass them in as keyword arguments, to minimize the
+                # impact on implementation signatures. Most vendor APIs
+                # will ignore one or more of these arguments.
+                fulfillment = api.fulfill(
+                    patron, pin, licensepool, internal_format=internal_format,
+                    part=part, fulfill_part_url=fulfill_part_url
+                )
+
+                if getattr(fulfillment, "can_cache_manifest", False):
+                    loan.cached_manifest = fulfillment.content
+                    loan.cached_content_type = fulfillment.content_type
```

Comment:
Some of the tests were failing since they had a `LicensePoolDeliveryMechanisms` that was acting as `fulfillment` here instead of a `FulfillmentInfo`. Due to that I wasn't sure if a `FulfillmentInfo` is always present.

Reply:
I'll look into this further and double check. I had some that were failing yesterday, but just changed this locally and it seems ok.

Reply:
I see... that sounds a little weird to me, so it could be a problem with the test or the code. Let me know what you find.

Reply:
Yeah, it looks like you exposed an error or shortcut in those tests that we can do without.

Reply:
I've updated these tests to use a `FulfillmentInfo` and replaced the `getattr()`.